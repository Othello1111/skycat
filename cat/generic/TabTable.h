// -*-c++-*-
#ifndef _TabTable_h_
#define _TabTable_h_

/*
 * E.S.O. - VLT project/ESO Archive
 * $Id: TabTable.h,v 1.1.1.1 2009/03/31 14:11:52 cguirao Exp $
 *
 * TabTable.h - class definitions for accessing values from a char buffer
 * 	        in the format of table, such as the result of a database query,
 *              where the rows are separated by newlines and the columns by
 *              tabs (or other given char).
 *
 * See the man page for a complete description.
 *
 * who             when       what
 * --------------  --------   ----------------------------------------
 * Allan Brighton  9 Jan 96  Created
 */


using namespace std;
#include <iostream>

/*
 * Class TabTable
 *
 * This class manages a buffer containing a table where the rows are
 * separated by newlines and the columns by tabs (or other char).  The
 * class provides transparent access to the table based on a row,column
 * index and allows for type conversion from string to the desired type.
 */
class TabTable {
protected:
    int numRows_;		// number of rows
    int numCols_;		// number of columns
    char** colNames_;		// array of column names

    char* buf_;			// saved copy of input buffer (memory for table)
    char** table_;		// array of row/col values
    int* index_;		// array of row indexes for sorting

    char sep_;			// separator char (default: \t)
    int status_;		// status after constructor (0 if OK)

    // these are set and used only during a table sort for use by compare methods
    static int numSortCols_;	// number of columns to sort by
    static char** sortCols_;	// array of sort column names
    static int* sortColIndexes_;// array of sort column indexes
    static int sortOrder_;	// sort order (increasing if >= 0)
    static int sortStatus_;	// status after sort (0 is ok)

    // scan the table to set colNames, numRows, numCols and set the start of data ptr
    virtual int scanTable(int maxRows, char*& start);

    // util to split a line into string array
    virtual int splitList(char* line, char** colValues);

    // report errors in input
    virtual int tab_error(int row, int col, char* expected, char* value) const;
   
    // scan the given buffer and return the number of lines
    virtual int getNumLines(char* buf, int maxRows);

    // create and fill the internal table from the given buffer in tab table format 
    virtual int fillTable(char* buf);

    // internal search util: assumes stream is positioned at first row
    virtual int search(istream& is, int numSearchCols, char** searchCols, 
	       char** minValues, char** maxValues, int maxRows);

    // given a row with columns for this table, compare the row with the
    // given column min and max values and return 0 if there is a match.
    virtual int compareRow(char* buf, int numSearchCols, char** searchCols, 
		   char** minValues, char** maxValues);

    // same as above, but with the row split into a string array
    virtual int compareRow(char** colValues, int numSearchmCols, char** searchCols, 
		   char** minValues, char** maxValues);

    int compareRow(const TabTable& table, int row, int numSearchCols, 
			 char** searchCols, char** minValues, char** maxValues);

    // return 0 if the given value is in the given range, doing a numeric comparison if
    // possible.
    virtual int compareCol(const char* value, const char* minValue, const char* maxValue);

    // return column index in input for the given column name
    // (may be redefined by a derived class taht changes column names or positions)
    virtual int inputColIndex(const char* colName) const {return colIndex(colName);}

    // check that row and column are in range
    virtual int checkTableIndex(int row, int col=0) const;

    // print table title and other info
    virtual void printTableTop(ostream& os, const char* title = NULL);

    // copy constructor (not defined)
    // TabTable(const TabTable&);

public:
    // constants 
    enum {MAX_ROW_SIZE=8*1024};	 // max size of a tab table row
    enum {MAX_HEADER_SIZE=1024}; // max size of a tab table heading
    enum {MAX_COLUMNS=255};	 // max number of table columns

public:
    // constructor: initialize empty table
    TabTable(char sep = '\t');

    // constructor: initialize table from buffer in tab table format
    TabTable(const char* buf, int maxRows = 0, char sep = '\t');

    // constructor: initialize table from data buffer without headings
    TabTable(int numCols, char** colNames, const char* buf, int maxRows = 0, char sep = '\t');

    // destructor: free any allocated memory
    virtual ~TabTable();

    // make the table empty and free any resources used
    virtual int clear();

    // fill the table from the given buffer in tab table format
    virtual int init(const char* buf, int maxRows = 0, int owner = 0);

    // fill the table from the given buffer in tab table format, with headings
    // specified separately
    virtual int init(int numCols, char** colNames, const char* buf, 
		     int maxRows = 0, int owner = 0);

    // access to row,column values
    // set the parameter value and return 0 on success
    virtual int get(int row, int col, char*& value) const;
    virtual int get(int row, int col, int& value) const;
    virtual int get(int row, int col, double& value) const;
    virtual int get(int row, int col, float& value) const;
    virtual int get(int row, int col, short& value) const;
    virtual int get(int row, int col, char& value) const;

    // get table values by column name
    virtual int get(int row, const char* colName, char*& value) const;
    virtual int get(int row, const char* colName, int& value) const;
    virtual int get(int row, const char* colName, double& value) const;
    virtual int get(int row, const char* colName, float& value) const;
    virtual int get(int row, const char* colName, short& value) const;
    virtual int get(int row, const char* colName, char& value) const;

    // return the table column index for the given table column name
    virtual int colIndex(const char* colName)  const;

    // return the column name for the given column index
    virtual const char* colName(int col)  const;

    // return true if the table contains the given column 
    virtual int hasCol(const char* name)  const {return (colIndex(name) >= 0);}

    // read the heading info from the given stream and return object for it
    static int head(const char* filename, TabTable&);
    static int head(istream&, TabTable&);
 
    // compare headings in this table and the given one
    virtual int compareHeadings(const TabTable& t);

    // compare the given rows
    virtual int compareRows(int row1, int row2);

    // save the contents of this object as a tab table
    virtual int save(const char* filename);
    virtual int save(ostream&);
    
    // append the contents of this object to the given tab table file
    virtual int append(const char* filename);
 
    // insert (or update) the contents of this object to the given tab table file
    virtual int insert(const char* filename, int col = 0);

    // remove rows in the tab table file that match the given col in this object
    virtual int remove(const char* filename, int col);
 
    // find a row in this object matching the col in the given tab separated table row
    virtual int findRow(const char* tableRow, int col);
    
    // sort the contents of this tab table by the given columns
    virtual int sort(int numSortCols, char** sortCols, int sortOrder = 0); 

    // Search the given tab table for upto maxRows rows with columns
    // values matching the given arguments and fill this table with the
    // resulting rows.
    virtual int search(const TabTable& table, int numSearchCols, char** searchCols, 
		       char** minValues, char** maxValues, int maxRows); 

    // search as above, but for a single column name and value
    virtual int search(const TabTable& table, const char* searchCol, const char* value, int maxRows);

    // search as above, but for a single column index and value
    virtual int search(const TabTable& table, int searchCol, const char* value, int maxRows); 

    // Search the given tab table file for upto maxRows rows with columns
    // values matching the given arguments and fill this table with the
    // resulting rows.
    virtual int search(const char* filename, int numSearchCols, char** searchCols, 
		       char** minValues, char** maxValues, int maxRows); 

    // search as above, but for a single column name and value
    virtual int search(const char* filename, const char* searchCol, const char* value, int maxRows);

    // search as above, but for a single column index and value
    virtual int search(const char* filename, int searchCol, const char* value, int maxRows); 

    // print the table rows to the given stream
    virtual int printRows(ostream& os) const;
    
    // print the given table row to the given stream
    virtual int printRow(ostream& os, int row) const; 

    // output operator
    friend ostream& operator<<(ostream& os, TabTable& t) {
	t.save(os); return os;
    }

    // member access
    virtual int numRows() const {return numRows_;}
    virtual void numRows(int n) {if (n < numRows_ && n >= 0) numRows_ = n;}
    virtual int numCols() const {return numCols_;}
    virtual char** colNames() const {return colNames_;}
    virtual int status() const {return status_;}
};



#endif /* _TabTable_h_ */
