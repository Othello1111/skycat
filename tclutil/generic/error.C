/*
 * E.S.O. - VLT project 
 * "@(#) $Id: error.C,v 1.2 1998/02/18 11:33:05 abrighto Exp $"
 *
 * error.C - error reporting routines
 * 
 * who             when      what
 * --------------  --------  ----------------------------------------
 * Allan Brighton  05/10/95  Created
 */
static const char* const rcsId="@(#) $Id: error.C,v 1.2 1998/02/18 11:33:05 abrighto Exp $";


#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <iostream.h>
#include <strstream.h>
#include <errno.h>
#include <stdio.h>
#include "error.h"

// static variable holding text of last error messages
static char errmsg_[5*1024];

// error code (see defines in <sys/errno.h> for values)
static int errno_ = 0;

// optional error handler, to be called with error messages
static void (*errhandler_)(const char*) = NULL;

// optional message handler, to be called with log messages
static void (*msghandler_)(const char*) = NULL;


/*
 * global error reporting routine
 */
int error(const char* msg1, const char* msg2, int code)
{
    char buf[sizeof(errmsg_)];
    ostrstream os(buf, sizeof(buf));
    os << msg1 << msg2 << ends;
    
    if (errhandler_)
	(*errhandler_)(buf);
    else
	print_error(buf);

#ifdef XXXDEBUG
    cerr << "debug: " << buf << endl;
#endif

    errno_ = code;
    strncpy(errmsg_, buf, sizeof(errmsg_)-1);
    return ERROR;
}


/*
 * report the error, including system error code
 */
int sys_error(const char* msg1, const char* msg2)
{
    extern int sys_nerr;
    extern char *sys_errlist[];
    extern int errno;
    char* s = strerror(errno);

    // XXX configure: sys_nerr may not be defined ?
    if (s == NULL || errno < 0 || errno >= sys_nerr)
	return error(msg1, msg2);

    char buf[sizeof(errmsg_)];
    ostrstream os(buf, sizeof(buf));
    os << msg1 << msg2 << ": " << (s ? s : sys_errlist[errno]) << ends;

    if (errhandler_)
	(*errhandler_)(buf);
    else
	print_error(buf);

#ifdef XXXDEBUG
    cerr << "debug: " << buf << endl;
#endif

    errno_ = errno;
    strcpy(errmsg_, buf);
    return ERROR;
}


/*
 * This routine has an interface like printf and reports an error in the
 * same way as error() above.
 */
int fmt_error(const char* fmt, ...)
{
    char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsprintf(buf, fmt, ap);
    va_end(ap);
    
    return error(buf);
}


/*
 * This routine has an interface like printf and reports a system error
 * in the same way as sys_error() above.
 */
int fmt_sys_error(const char* fmt, ...)
{
    char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsprintf(buf, fmt, ap);
    va_end(ap);
    
    return sys_error(buf);
}


/* 
 * return the text of the previous error message
 */
char* last_error() 
{
    return errmsg_;
}


/* 
 * return the error code for the previous error
 */
int last_error_code() 
{
    return errno_;
}


/* 
 * reset the last_error buf to empty 
 */
void clear_error()
{
    errmsg_[0] = '\0';
}


/*
 * set a routine to be called with the text of error messages
 * when they occur. The argument is a pointer to an error 
 * handler:
 *
 *       void errhandler(const char* msg);
 *
 * The return value is a pointer to the previous error handler, or NULL,
 * if none was defined.
 */
void (*set_error_handler(void (*errhandler)(const char*)))(const char*)
{
    void (*old_handler)(const char*) = errhandler_;
    errhandler_ = errhandler;
    return old_handler;
}


/*  
 * print the given message on stderr (may be used as an error
 * handler)
 */
void print_error(const char* msg)
{
    fprintf(stderr, "%s\n", msg);
    fflush(stderr);
}


/* ---- The routines below are for log messages that are not errors --- */



/*
 * This routine logs a message and has an interface like printf.
 */
void log_message(const char* fmt, ...)
{
    char buf[1024];
    va_list ap;
    va_start(ap, fmt);
    vsprintf(buf, fmt, ap);
    va_end(ap);
    
    if (msghandler_)
	(*msghandler_)(buf);
    else 
	print_log_message(buf);
}

/*
 * set a routine to be called with the text of log messages
 * when they are made. The argument is a pointer to a log message 
 * handler:
 *
 *       void msghandler(const char* msg);
 *
 * The return value is a pointer to the previous message handler, or NULL,
 * if none was defined.
 */
void (*set_log_message_handler(void (*msghandler)(const char*)))(const char*)
{
    void (*old_handler)(const char*) = msghandler_;
    msghandler_ = msghandler;
    return old_handler;
}


/*  
 * print the given message on stdout (may be used as a message
 * handler)
 */
void print_log_message(const char* msg)
{
    printf("%s\n", msg);
    fflush(stdout);
}

