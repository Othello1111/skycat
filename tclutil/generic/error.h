#ifndef _error_h_
#define _error_h_
/*
 * E.S.O. - VLT project 
 * "@(#) $Id: error.h,v 1.2 2005/02/02 01:43:01 brighton Exp $" 
 *
 * error.h - declarations for global error reporting
 * 
 * who             when      what
 * --------------  --------  ----------------------------------------
 * Allan Brighton  05/10/95  Created
 */

/* error status */
#undef OK
#undef ERROR
enum {OK, ERROR};

/* report the error */
int error(const char* msg1, const char* msg2="", int code = 0);

/* report the error with the system error code, like perror */
int sys_error(const char* msg1, const char* msg2="");

/* report error with a printf style interface */
int fmt_error(const char* fmt, ...);

/* report a system error with a printf style interface */
int fmt_sys_error(const char* fmt, ...);

/* return the text of the previous error message */
char* last_error();

/* return the error code for the previous error */
int last_error_code();

/* reset the last_error buf to empty */
void clear_error();

/* set a routine to be called when errors occur and return ptr to previous handler */
void (*set_error_handler(void (*error_handler)(const char*)))(const char*);

/* simple error handler: print the message */
void print_error(const char* msg);

/* print a message (like printf) */
void log_message(const char* fmt, ...);

/* set a routine to be called for log messages and return ptr to previous handler */
void (*set_log_message_handler(void (*message_handler)(const char*)))(const char*);

/* simple message handler: print the message */
void print_log_message(const char* msg);



#endif /* _error_h_ */

