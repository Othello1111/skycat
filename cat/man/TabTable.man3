# E.S.O. - VLT project/ESO Archive
# $Id: TabTable.man3,v 1.1.1.1 2006/01/12 16:36:07 abrighto Exp $
#
# This file is processed by the ESO/VLT docDoManPages command to
# produce a man page in nroff, TeX and MIF formats. 
# See docDoManPages(1) for a description of the input format.
# 
# who             when       what
# --------------  ---------  ----------------------------------------
# Allan Brighton  27 Jun 96  Created
#

NAME
TabTable - Class for managing a tab separated table of values

SYNOPSIS
#include "TabTable.h"
class TabTable {
...
public:
    // constants 
    enum {MAX_ROW_SIZE=8*1024};	 // max size of a tab table row
    enum {MAX_HEADER_SIZE=1024}; // max size of a tab table heading
    enum {MAX_COLUMNS=255};	 // max number of table columns

public:
    TabTable(char sep = '\t');
    TabTable(const char* buf, int maxRows = 0, char sep = '\t');
    TabTable(int numCols, char** colNames, const char* buf, 
             int maxRows = 0, char sep = '\t');
    virtual ~TabTable();
    
    virtual int clear();

    virtual int init(const char* buf, int maxRows = 0);
    virtual int init(int numCols, char** colNames, const char* buf, 
                     int maxRows = 0);

    virtual int get(int row, int col, char*& value);
    virtual int get(int row, int col, int& value);
    virtual int get(int row, int col, double& value);
    virtual int get(int row, int col, float& value);
    virtual int get(int row, int col, short& value);
    virtual int get(int row, int col, char& value);

    virtual int get(int row, const char* colName, char*& value);
    virtual int get(int row, const char* colName, int& value);
    virtual int get(int row, const char* colName, double& value);
    virtual int get(int row, const char* colName, float& value);
    virtual int get(int row, const char* colName, short& value);
    virtual int get(int row, const char* colName, char& value);

    virtual int colIndex(const char* colName)  const;
    virtual const char* colName(int col)  const;
    virtual int hasCol(const char* name)  const;

    static int head(const char* filename, TabTable&);
    static int head(istream&, TabTable&);

    virtual int compareHeadings(const TabTable& t);

    virtual int save(const char* filename);
    virtual int save(ostream&);

    virtual int append(const char* filename);
    virtual int remove(const char* filename);

    virtual int findRow(const char* tableRow);
    virtual int printRows(ostream& os);
    
    virtual int search(const char* filename, int numSearchCols, 
                       char** searchCols, 
		       char** minValues, char** maxValues, 
                       int maxRows); 

    virtual int search(const char* filename, const char* searchCol, 
                       const char* value, int maxRows);

    virtual int search(const char* filename, int searchCol, 
                       const char* value, int maxRows); 

    friend ostream& operator<<(ostream& os, TabTable& t);
    
    virtual int numRows() const;
    virtual void numRows(int n);
    virtual int numCols() const;
    virtual char** colNames() const;
    virtual int status() const;
};


DESCRIPTION
Class TabTable manages a buffer containing a table where the rows are
separated by newlines and the columns by tabs (or other specified
char).  The class provides transparent access to the table based on a
row,column index and allows for type conversion from string to the
desired type. Methods are also available to print the table and save
it to a file.

PUBLIC METHODS

TabTable(char sep = '\t')
	Constructor: initialize empty table.

TabTable(const char* buf, int maxRows = 0, char sep = '\t')
	Constructor: initialize table from buffer in tab table format: 

	VAR	VALUE	VALUE	...
	...
	COL1	COL2	COL3	...	COLN
	---		----	----	----
	data	data	data		data
	...

	Currently, anything up to the column headings is ignored (This
	may change later). The format is the same as the tab table
	format used by starbase.  A line beginning with a "-"
	separates the column names from the data and all column
	headings and data are separated by tabs (or sep char).

	The part preceding the column headings may be used later to
	set certain variables pertaining to the table. In this case, a
	variable may have one or more values, separated by tabs (sep
	char).

	If maxRows is nonzero, only upto that many rows are taken from
	buf.

TabTable(int numCols, char** colNames, const char* buf, int maxRows = 0, char sep = '\t')
	Constructor: initialize table from data buffer without If
	maxRows is nonzero, only upto that many rows are taken from
	buf.  headings. The first two args specify the number column
	headings and their names.

~TabTable()
	Destructor: free any allocated memory

init(const char* buf, int maxRows = 0)
	Fill the table from the given buffer in tab table format.  If
	maxRows is nonzero, only upto that many rows are taken from
	buf.

init(int numCols, char** colNames, const char* buf, int maxRows = 0)
	Fill the table from the given buffer in tab table format, with
	headings specified separately.  The first two args specify the
	number column headings and their names.  If maxRows is
	nonzero, only upto that many rows are taken from buf.

clear()
	Make the table empty and free any resources used

get(int row, int col, char*& value)
get(int row, int col, int& value)
get(int row, int col, double& value)
get(int row, int col, float& value)
get(int row, int col, short& value)
get(int row, int col, char& value)
	Get row,column values: the parameter "value" is set and 0 is
	returned if there are no errors. Values of type char* point to
	internal storage and should not be modified.

get(int row, const char* colName, char*& value)
get(int row, const char* colName, int& value)
get(int row, const char* colName, double& value)
get(int row, const char* colName, float& value)
get(int row, const char* colName, short& value)
get(int row, const char* colName, char& value)
	Get table values by column name. Values of type char* point to
	internal storage and should not be modified.

colIndex(const char* colName)  const
	Return the table column index for the given table column name.

const char* colName(int col)  const
	Return the column name for the given column index.

hasCol(const char* name)  const
	Return true if the table contains the given column. 

static int head(const char* filename, TabTable&)
static int head(istream&, TabTable&)
 	Read the heading info from the given stream and put it in the
 	given table.

compareHeadings(const TabTable& t)
	Compare headings in this table and the given one and return 0
	if they are the same.

save(const char* filename)
save(ostream&)
	Save the contents of this object to the given file as a tab
	table.

append(const char* filename)
 	Append the contents of this object to the given tab table file

remove(const char* filename)
 	Remove rows in the tab table file that are also in this object.

findRow(const char* tableRow)
 	Find a row matching the given tab separated one and return the
 	index.

search(const char* filename, int numSearchCols, char** searchCols, 
       char** minValues, char** maxValues, int maxRows)
 
	Search the given tab table file for upto maxRows rows with
	columns values matching the given arguments and fill this
	table with the resulting rows.

	Args:

	filename          - tab table file to search
	numSearchCols     - number of columns in argument arrays

	searchCols        - names of the columns to compare

	minValues         - min/max values for comparison
	maxValues

	maxRows           - max number of rows to find


search(const char* filename, const char* searchCol, const char* value, 
       int maxRows);

	Search the given tab table file for upto maxRows rows where
	the given column has the given value and fill this table with
	the resulting rows.

	Args:

	filename         - tab table file to search

	searchCol        - name of the column to compare

	value            - value for comparison

	maxRows          - max number of rows to find


search(const char* filename, int searchCol, const char* value, 
       int maxRows); 

	Search the given tab table file for upto maxRows rows where
	the given column has the given value and fill this table with
	the resulting rows.

	Args:

	filename         - tab table file to search

	searchCol        - index of the column to compare (0..n)

	value            - value for comparison

	maxRows          - max number of rows to find


printRows(ostream& os)
	Print the table rows to the given stream in tab table format.

friend ostream& operator<<(ostream& os, TabTable& t)
	Output operator, prints the table to the given stream in tab
	table format.

numRows() const
	Return the number of rows in the table.

void numRows(int n)
	Set the number of rows in the table.

numCols() const
	Return the number of columns in the table.

colNames() const
	Return an array of column names in the table.

status() const
	Return the status after the constructor has completed.

SEE ALSO

QueryResult
----------------------------------------------------------------------
