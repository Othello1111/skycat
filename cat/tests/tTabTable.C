/*
 * E.S.O. - VLT project 
 * $Id: tTabTable.C,v 1.1.1.1 2006/01/12 16:36:09 abrighto Exp $
 *
 * tTabTable.C - test cases for class TabTable
 * 
 * who             when       what
 * --------------  --------   ----------------------------------------
 * Allan Brighton  08 Jan 96  Created
 */

using namespace std;
#include <stdio.h>
#include <stdlib.h>
#include "error.h"
#include "TabTable.h"

main()  
{
    // gcc doesn't need this, but SunCC does...
    ios::sync_with_stdio();

    // errors will be printed on stderr automatically
    set_error_handler(print_error);

    // dummy table data in Tab list format
    char buf[256];
    sprintf(buf, "col1\tcol2\tcol3\n---\t---\t---\n%s\n%s\n%s\n%s\n",
	    "1\t1.0\taa", 
	    "2\t2.0\tbbbb", 
	    "3\t3.0\tc c c", 
	    "4\t4.0\t");
    TabTable t(buf);
    if (t.status() != 0) {
	printf("test 1 failed\n");
	exit(1);
    }

    int n;
    double d;
    float f;
    short s;

    for (int i = 0; i < t.numRows(); i++) {
	for (int j = 0; j < t.numCols(); j++) {
	    char* s;
	    if (t.get(i, j, s) != 0) {
		printf("test 2 failed at row %d, col %d\n", i, j);
		exit(1);
	    }
	    printf("row %d: %s =  \"%s\"\n", i, t.colName(j), s);
	}
	if (t.get(i, 0, n) != 0) {
	    printf("test 3 failed at row %d, col 0\n", i);
	    exit(1);
	}
	if (t.get(i, 0, s) != 0) {
	    printf("test 4 failed at row %d, col 0\n", i);
	    exit(1);
	}
	if (t.get(i, 0, d) != 0) {
	    printf("test 5 failed at row %d, col 0\n", i);
	    exit(1);
	}
	if (t.get(i, 1, d) != 0) {
	    printf("test 6 failed at row %d, col 1\n", i);
	    exit(1);
	}
	if (t.get(i, 1, f) != 0) {
	    printf("test 7 failed at row %d, col 1\n", i);
	    exit(1);
	}
    }
    if (n != 4 || s != 4 || d != 4.0 || f != 4.0) {
	printf("test 8 failed\n");
	exit(1);
    }

    // test save method
    char* filename = "tTabTable.out";
    if (t.save(filename) != 0) {
	printf("test 9 failed\n");
	exit(1);
    }

    // test append method
    if (t.append(filename) != 0) {
	printf("test 10 failed\n");
	exit(1);
    }

    // test alternate constructor
    TabTable* t2 = new TabTable(t.numCols(), t.colNames(), buf);
    if (t.compareHeadings(*t2)) {
	printf("test 11 failed\n");
	exit(1);
    }
    TabTable* t3 = new TabTable(t2->numCols(), t2->colNames(), buf);
    delete t2;
    TabTable t4;
    if (TabTable::head(filename, t4) != 0) {
	printf("test 12 failed\n");
	exit(1);
    }
    if (t3->compareHeadings(t4)) {
	printf("test 13 failed\n");
	exit(1);
    } 

    // test search method on file saved above
    cout << "testing search method:\n";
    TabTable t5;
    char** searchCols = t.colNames();
    int numCols = t.numCols();
    char* minValues[3];
    char* maxValues[3];
    minValues[0] = "2";		// col0
    maxValues[0] = "3";	
    minValues[1] = "1";		// col1
    maxValues[1] = "3";	
    minValues[2] = "a";		// col2
    maxValues[2] = "d";	
    if (t5.search(filename, numCols, searchCols, 
		  minValues, maxValues, 10) != 0) {
	printf("test 14 failed\n");
	exit(1);
    }
    cout << "results of search:\n" << t5;

    // test sort method
    int ncols = 1;
    char* colName = "col2";
    if (t5.sort(1, &colName) != 0) {
	printf("test 15 failed\n");
	exit(1);
    }
    cout << "results after sort:\n" << t5;

    // test search for single value
    cout << "\ntesting search method for single column:\n";
    if (t5.search(filename, "col2", "2.0", 1) != 0) {
	printf("test 16 failed\n");
	exit(1);
    }
    cout << "results of search:\n" << t5;
	
    return 0;
}
