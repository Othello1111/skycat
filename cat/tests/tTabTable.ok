row 0: col1 =  "1"
row 0: col2 =  "1.0"
row 0: col3 =  "aa"
row 1: col1 =  "2"
row 1: col2 =  "2.0"
row 1: col3 =  "bbbb"
row 2: col1 =  "3"
row 2: col2 =  "3.0"
row 2: col3 =  "c c c"
row 3: col1 =  "4"
row 3: col2 =  "4.0"
row 3: col3 =  ""
testing search method:
results of search:
TabTable
col1	col2	col3
----	----	----
2	2.0	bbbb
3	3.0	c c c
2	2.0	bbbb
3	3.0	c c c
results after sort:
TabTable
col1	col2	col3
----	----	----
2	2.0	bbbb
2	2.0	bbbb
3	3.0	c c c
3	3.0	c c c

testing search method for single column:
results of search:
TabTable
col1	col2	col3
----	----	----
2	2.0	bbbb
