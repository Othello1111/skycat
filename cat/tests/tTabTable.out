TabTable
col1	col2	col3
----	----	----
1	1.0	aa
2	2.0	bbbb
3	3.0	c c c
4	4.0	
1	1.0	aa
2	2.0	bbbb
3	3.0	c c c
4	4.0	
